### ISSUE TITLE
[css-logical] Should the mapping for logical values depend on the element or containing block?

### ISSUE THREAD

**Issue by Loirooriol** [w3c/csswg-drafts#3013]
https://drafts.csswg.org/css-logical/#float-clear says

> The mapping on these properties uses the writing mode of the element’s containing block.

However, on Firefox the mapping depends on the writing mode of the element itself. I didn't read the spec carefully when implementing logical values in Blink, so I ended up copying the Firefox behavior. https://github.com/web-platform-tests/wpt/pull/12326 tests the implemented behavior and not the spec one.

Instead of fixing both implementations and WPT, maybe consider changing the spec?

This would be consistent with the previous change that made the mapping for logical properties depend on the writing mode of the element instead of the one of the parent element.
https://github.com/w3c/csswg-drafts/blob/206498e6bd16ad8dc00967a26607c72bdcd12dfa/css-logical-1/Overview.bs#L731

**Comment by emilio** [comment]
The containing block is a layout concept, so the style engine doesn't necessarily have that information. I'd prefer not to teach the style system about containing blocks, I think the spec should be changed here.

**Comment by fantasai** [comment]
We have an outstanding resolution on this, I guess it just didn't make it into the spec (which is has in general been quite neglected). I'll be spending some time on it the next few days, though, so outstanding resolutions should get edited in soonish. :)

**Comment by fantasai** [comment]
Oh, sorry, thought this was about the logical properties. Those compute against the element's own writing mode always, for simplicity and because it wasn't entirely clear where to draw the boundary between “this is best computed against the CB writing mode” and “this is best computed against the element's own writing mode”: at the border edge? the padding edge? the content edge?

For logical values, it does depend on the property whether the containing block or the element itself is referenced. For example, the `start` and `end` keywords on `align-self` (which aligns the element in its containing block) use the containing block, so that when you set alignment on a bunch of items they are consistently aligned regardless of writing mode. The `self-start` and `self-end` values allow to reference the element itself. However, `text-align` and `align-content`, which align an element’s content within itself, reference the element's own writing mode, since the containing block's writing mode would be irrelevant.

Floats are more similar to `align-self` and `justify-self`, so I think they should compute against the CB. Typically you don't want the float to float differently based on its own contents, but rather on the  writing mode of its context.

@emilio Logical values don't compute through to physical ones, so the style system doesn't need to know about it.

**Comment by Loirooriol** [comment]
> Logical values don't compute through to physical ones

That's not clear: #2821

**Comment by fantasai** [comment]
@Loirooriol Yes it is, the propdef tables for those properties say “Computed value: as specified”. That means the values do not compute through. This is important because (as you point out) it would break inherited properties to do otherwise.

**Comment by FremyCompany** [comment]
I agree with @fantasai here:
1. if you specify "start" your children should inherit "start", as the style system doesn't care about containing block
2. however what "start" means sometimes can only be understood in the context of the containing block when you are doing layout, so its value should be preserved untouched until then.

It is kinda unfortunate implementations did something else here,

**Comment by Loirooriol** [comment]
> the propdef tables for those properties say “Computed value: as specified”

@fantasai But the propdef tables for those properties don't include the logical values, so it's not clear. The partial propdef table in CSS Logical with the new values should repeat “Computed value: as specified”

**Comment by fantasai** [comment]
@Loirooriol Fair, fixed in https://github.com/w3c/csswg-drafts/commit/f3953ed6c245df059df80034c427491b5433bc5f for #2821.
